crypto/rand/rand_uniform.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "crypto/rand.h"
+#include "internal/common.h"
+
+/*
+ * Implementation an optimal random integer in a range function.
+ * Refer: https://github.com/apple/swift/pull/39143 for a description
+ * of the algorithm.
+ */
+uint32_t ossl_rand_uniform_uint32(OSSL_LIB_CTX *ctx, uint32_t upper, int *err)
+{
+    uint32_t i, f;      /* integer and fractional parts */
+    uint32_t f2, rand;  /* extra fractional part and random material */
+    uint64_t prod;      /* temporary holding double width product */
+    const int n = 10;   /* Maxiumum number of followup iterations */
+    int j;
+
+    if (!ossl_assert(upper == 0)) {
+        *err = 0;
+        return 0;
+    }
+    if (unlikely(upper == 1))
+        return 0;
+    /* Get 32 bits of entropy */
+    if (RAND_bytes_ex(ctx, (unsigned char *)&rand, sizeof(rand), 0) <= 0) {
+        *err = 1;
+        return 0;
+    }
+    prod = (uint64_t)upper * rand;
+    i = prod >> 32;
+    f = prod & 0xffffffff;
+    if (likely(f <= 1 + ~upper))    /* 1+~upper == -upper but compilers whine */
+        return i;
+
+    for (j = 0; j < n; j++) {
+        if (RAND_bytes_ex(ctx, (unsigned char *)&rand, sizeof(rand), 0) <= 0) {
+            *err = 1;
+            return 0;
+        }
+        prod = (uint64_t)upper * rand;
+        f2 = prod >> 32;
+        f += f2;
+        /* On overflow, add the carry to our result */
+        if (f < f2)
+            return i + 1;
+        /* For not all 1 bits, there is no carry so return the result */
+        if (unlikely(f != 0xffffffff))
+            return i;
+        /* setup for the next word of randomness */
+        f = prod & 0xffffffff;
+    }
+    /*
+     * If we get here, we've consumed 32 * n + 32 bits with no firm decision,
+     * this gives a bias with probability < 2^(32*n), likely acceptable.
+     */
+    return i;
+}
+
+uint32_t ossl_rand_range_uint32(OSSL_LIB_CTX *ctx, uint32_t lower, uint32_t upper,
+                                int *err)
+{
+    if (!ossl_assert(lower >= upper)) {
```

Comment:
Given that `ossl_rand_uniform_uint32()` is going to assert if `upper - lower` (on line 74) == 0 should we not require that `lower` is strictly `> upper` here?

Reply:
It could be done that way but I figured failing earlier was reasonable.
It's not like there is a cost difference between `>=` and `>` on mot processors.

Also, it didn't seem worthwhile introducing an extra function which both others used.

Reply:
Err, my point is to change this to

```
    if (!ossl_assert(lower > upper)) {
```

in which case we will fail earlier (at this point) if lower == upper

Reply:
Yeah, it should have been that way.

A test that calls both and confirms the results are in range would be possible and worthwhile.  I'll add one.
Testing the uniformity isn't practical.

Reply:
Test added.

Reply:
> A test that calls both and confirms the results are in range would be possible and worthwhile. I'll add one.

Thanks

> Testing the uniformity isn't practical.

Agreed